packages/keyring-controller/src/KeyringController.ts

```diff
@@ -1015,18 +1028,20 @@ export class KeyringController extends BaseController<
    * @returns Promise resolving to current state.
    */
   async setLocked(): Promise<KeyringControllerMemState> {
-    this.#unsubscribeFromQRKeyringsEvents();
+    return this.#withControllerLock(async () => {
+      this.#unsubscribeFromQRKeyringsEvents();
 
-    this.#password = undefined;
-    this.update((state) => {
-      state.isUnlocked = false;
-      state.keyrings = [];
-    });
-    await this.#clearKeyrings();
+      this.#password = undefined;
+      this.update((state) => {
+        state.isUnlocked = false;
+        state.keyrings = [];
+      });
+      await this.#clearKeyrings();
 
-    this.messagingSystem.publish(`${name}:lock`);
+      this.messagingSystem.publish(`${name}:lock`);
```

Comment:
I notice that for `removeAccount`, the `accountRemoved` is emitted after the lock is released, but in this case the `lock` event is emitted before the lock is released. Do these need to behave the same or does it matter?

Reply:
Good point! I moved that out of the lock in `removeAccount` for two reasons mainly:
1. I initially had a deadlock concern for side effects running while the mutex was still locked, but this has been proven not to be the case
2. In the subsequent PR for [atomic operations](https://github.com/MetaMask/core/pull/4192/files#diff-cacd62f731d7f6947fe841f6ba3441c7a16f764b699801b0c212e0e6c15879bfR977) this method will be wrapped in `this.#persistOrRollback`, and since the state and vault update will be delegated to the wrapper, this function would emit the `:accountRemoved` event before removing the account from the state, which would not make sense
3. (Related to (2)) `setLocked` does not persist to the vault, it just clears the keyrings array, so it will not be wrapped in the `this.#persistOrRollback` function